### DISCUSSION TITLE
Multiple uri's for a single Pointer in jsonschema_frame.

### DISCUSSION THREAD

**Discussion by suprith-hub** [sourcemeta/jsontoolkit#567] [Q&A] [answered]
This [line](https://github.com/sourcemeta/jsontoolkit/blob/c834d475fcfc1c81b43eda02a814f44c648d2cc3/src/jsonschema/reference.cc#L177) iterates on the nearest bases for a subschema, on multiple bases obtained from [this](https://github.com/sourcemeta/jsontoolkit/blob/c834d475fcfc1c81b43eda02a814f44c648d2cc3/src/jsonschema/reference.cc#L121).
My question is: Why do we need to store multiple URI's for a single location - i.e, Pointer here

**Comment by jviotti** [comment]
A single location can be identified using multiple URIs. For example, if you have a schema resource that has its own `$id` that embeds another JSON Schema resource that also has its own `$id`, you can reference to any part of the nested schema resource relative to any of the base URIs. Also, if a subschema has an anchor, then you can also reference that same location using an anchor.

In summary, there is no requirement for a schema location to be identified by a single URI. In many cases, there will be multiple URIs to a single schema location!

**Reply by suprith-hub** [reply]
Okay that means base_uri is storing URI's with fragment even?.
 So, [this](https://github.com/sourcemeta/jsontoolkit/blob/c834d475fcfc1c81b43eda02a814f44c648d2cc3/src/jsonschema/reference.cc#L161) sets the absoluteURI right and if not present its nullptr? 
Even [here](https://github.com/sourcemeta/jsontoolkit/blob/c834d475fcfc1c81b43eda02a814f44c648d2cc3/src/jsonschema/reference.cc#L182) I feel the relative id's like $id:"/foo" are resolved against base uri's by subpointer walker.  
So, I didnt get it completely

**Reply by jviotti** [reply, marked as answer]
The base URIs could indeed contain fragments. It depends on the schema. Its just the URI (or URIs) that we resolve potentially relative URIs against in any given context. 

In any case, framing is tricky scary code 😅 If you have a specific schema you are studying, I'd recommend going through the framing code using a step debugger like LLDB so you see each intermediate calculation. If you are on Windows, opening JSON Toolkit on Visual Studio Community (not Visual Studio Code) would be a good way of using its debugger: https://learn.microsoft.com/en-us/visualstudio/debugger/debugger-feature-tour?view=vs-2022

I don't remember from the top of my head (I'm an Apple guy :) ), but I think Visual Studio supports CMake out of the box. Otherwise, you can use CMake's Visual Studio generator to output a Visual Studio solution that you can open in the IDE: https://stackoverflow.com/questions/395169/using-cmake-to-generate-visual-studio-c-project-files

**Reply by jviotti** [reply]
Other than, I'd also recommend carefully studying the JSON Schema referencing suite by Julian Berman (https://github.com/python-jsonschema/referencing-suite/tree/main/tests/json-schema-draft-2020-12). My framing code is tested with that suite. It is a good digestable way of understanding how schemas and subschemas are meant to be identified

**Reply by jviotti** [reply]
Also, at some point, @RonItelman and I would love to release a desktop app tool with a nice GUI to better understand framing and other tricky aspects of JSON Schema for educational/debugging purposes
